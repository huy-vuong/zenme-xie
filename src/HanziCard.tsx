import PropTypes from 'prop-types';
import React from 'react';
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import styles from 'zenme-xie/HanziCard.module.scss';
import HanziGrid from 'zenme-xie/HanziGrid';
import HanziSteps from 'zenme-xie/HanziSteps';

export default function HanziCard({ character, index }: HanziCardProp) {
  return (
    <Card className={styles.hanziCard}>
      <Card.Body>
        <Card.Title>{index + 1}</Card.Title>
        <Card.Title>{character}</Card.Title>
        <Container>
          <Row className="justify-content-md-center">
            <Col lg={4} md={12}>
              <div className={styles.hanziGrid}>
                {
                  <HanziGrid
                    character={character}
                    id={`hanzi-grid-${index}-${character.charCodeAt(0)}`}
                    delayBetweenStrokes={250}
                    radicalColor="#dc3545"
                    animateOnClick={true}
                  />
                }
              </div>
            </Col>
            <Col lg={8} md={12}>
              <div className="d-xl-none d-lg-none">
                <br />
              </div>
              <HanziSteps
                character={character}
                id={`hanzi-steps-${index}-${character.charCodeAt(0)}`}
                activeStrokeColor="#dc3545"
              />
            </Col>
          </Row>
        </Container>
        <br />
        <br />
        <br />
      </Card.Body>
    </Card>
  );
}

interface HanziCardProp {
  character: string;
  index: number;
}

HanziCard.propTypes = {
  character: PropTypes.string.isRequired,
  index: PropTypes.number.isRequired,
};
