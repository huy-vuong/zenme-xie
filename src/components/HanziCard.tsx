import PropTypes from 'prop-types';
import React from 'react';
import Card from 'react-bootstrap/Card';
import Col from 'react-bootstrap/Col';
import Container from 'react-bootstrap/Container';
import Row from 'react-bootstrap/Row';
import styles from 'zenme-xie/components/HanziCard.module.scss';
import HanziDefinition from 'zenme-xie/components/HanziDefinition';
import HanziGrid from 'zenme-xie/components/HanziGrid';
import HanziSteps from 'zenme-xie/components/HanziSteps';
import PinyinList from 'zenme-xie/components/PinyinList';
import HanziDictionaryEntry from 'zenme-xie/types/HanziDictionaryEntry';
import HanziGraphicsEntry from 'zenme-xie/types/HanziGraphicsEntry';

export default function HanziCard({
  character,
  dictionaryEntry,
  graphicsEntry,
  index,
}: HanziCardProp) {
  return (
    <Card className={styles.hanziCard}>
      <Card.Body>
        <Card.Title>{index + 1}</Card.Title>
        <Card.Title>{character}</Card.Title>
        <Container>
          <Row className="justify-content-md-center">
            <Col lg={4} md={12}>
              <div
                className={styles.hanziGrid}
                key={`hanzi-grid-${index}-${character.charCodeAt(0)}`}
              >
                <HanziGrid
                  character={character}
                  id={`hanzi-grid-${index}-${character.charCodeAt(0)}`}
                  delayBetweenStrokes={250}
                  radicalColor="#dc3545"
                  animateOnClick={true}
                />
              </div>
              {dictionaryEntry?.pinyin ? <br /> : null}
              <PinyinList
                id={`pinyin-list-${index}-${character.charCodeAt(0)}`}
                pinyinList={dictionaryEntry?.pinyin}
              />
              {dictionaryEntry?.definition ? <br /> : null}
              <HanziDefinition definition={dictionaryEntry?.definition} />
            </Col>
            <Col
              lg={8}
              md={12}
              key={`col-hanzi-steps-${index}-${character.charCodeAt(0)}`}
            >
              <div className="d-xl-none d-lg-none">
                <br />
              </div>
              <HanziSteps
                graphicsEntry={graphicsEntry}
                id={`hanzi-steps-${index}-${character.charCodeAt(0)}`}
                activeStrokeColor="#dc3545"
              />
            </Col>
          </Row>
        </Container>
        <br />
        <br />
        <br />
      </Card.Body>
    </Card>
  );
}

interface HanziCardProp {
  character: string;
  dictionaryEntry: HanziDictionaryEntry | null;
  graphicsEntry: HanziGraphicsEntry | null;
  index: number;
}

HanziCard.propTypes = {
  character: PropTypes.string.isRequired,
  dictionaryEntry: PropTypes.shape({
    character: PropTypes.string,
    definition: PropTypes.string,
    pinyin: PropTypes.arrayOf(PropTypes.string),
  }),
  graphicsEntry: PropTypes.shape({
    character: PropTypes.string,
    strokes: PropTypes.arrayOf(PropTypes.string),
    medians: PropTypes.arrayOf(
      PropTypes.arrayOf(PropTypes.arrayOf(PropTypes.number))
    ),
    radStrokes: PropTypes.arrayOf(PropTypes.number),
  }),
  index: PropTypes.number.isRequired,
};
