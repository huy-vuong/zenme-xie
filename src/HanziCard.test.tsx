import { render } from '@testing-library/react';
import React from 'react';
import HanziCard from 'zenme-xie/HanziCard';

test('renders a card with the given character printed', () => {
  const { getByText } = render(<HanziCard character="貓" index={0} />);
  getByText('貓');
});

test('renders a card with its ordinal number', () => {
  const { getByText } = render(<HanziCard character="貓" index={0} />);
  getByText('1');
});

test('renders a card containing the character in a rice grid', () => {
  const { asFragment } = render(<HanziCard character="貓" index={0} />);
  expect(asFragment()).toMatchSnapshot();
});
