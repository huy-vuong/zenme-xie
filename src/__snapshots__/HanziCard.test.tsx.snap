// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a card containing the character in a rice grid 1`] = `
<DocumentFragment>
  <div
    class="hanzi-card card"
  >
    <div
      class="card-body"
    >
      <div
        class="card-title h5"
      >
        1
      </div>
      <div
        class="card-title h5"
      >
        è²“
      </div>
      <div
        class="container"
      >
        <div
          class="justify-content-md-center row"
        >
          <div
            class="col-lg-4 col-md-12"
          >
            <div
              class="hanzi-animation"
            >
              <svg
                height="300"
                id="hanzi-grid-0-35987"
                width="300"
                xmlns="http://www.w3.org/2000/svg"
              >
                <line
                  class="rice-grid-guide"
                  data-testid="rice-grid-guide-nw-se"
                  x1="0"
                  x2="300"
                  y1="0"
                  y2="300"
                />
                <line
                  class="rice-grid-guide"
                  data-testid="rice-grid-guide-ne-sw"
                  x1="300"
                  x2="0"
                  y1="0"
                  y2="300"
                />
                <line
                  class="rice-grid-guide"
                  data-testid="rice-grid-guide-nn-ss"
                  x1="150"
                  x2="150"
                  y1="0"
                  y2="300"
                />
                <line
                  class="rice-grid-guide"
                  data-testid="rice-grid-guide-ww-ee"
                  x1="0"
                  x2="300"
                  y1="150"
                  y2="150"
                />
                <line
                  class="rice-grid-border"
                  data-testid="rice-grid-border-nw-ne"
                  x1="0"
                  x2="300"
                  y1="0"
                  y2="0"
                />
                <line
                  class="rice-grid-border"
                  data-testid="rice-grid-border-nw-sw"
                  x1="0"
                  x2="0"
                  y1="0"
                  y2="300"
                />
                <line
                  class="rice-grid-border"
                  data-testid="rice-grid-border-ne-se"
                  x1="300"
                  x2="300"
                  y1="0"
                  y2="300"
                />
                <line
                  class="rice-grid-border"
                  data-testid="rice-grid-border-sw-se"
                  x1="0"
                  x2="300"
                  y1="300"
                  y2="300"
                />
                <defs />
              </svg>
            </div>
          </div>
          <div
            class="col-lg-8 col-md-12"
          >
            <div
              class="d-xl-none d-lg-none"
            >
              <br />
            </div>
            <div
              class="hanzi-steps"
              id="hanzi-steps-0-35987"
            />
          </div>
        </div>
      </div>
      <br />
      <br />
      <br />
    </div>
  </div>
</DocumentFragment>
`;
